config/v1alpha1/types_image_policy.go

```diff
@@ -0,0 +1,228 @@
+// +kubebuilder:validation:XValidation:rule="has(self.policyType) && self.policyType == 'FulcioCAWithRekor' ? has(self.fulcioCAWithRekor) : !has(self.fulcioCAWithRekor)",message="fulcioCAWithRekor is required when policyType is FulcioCAWithRekor, and forbidden otherwise"
+type PolicyRootOfTrust struct {
+	// policyType serves as the union's discriminator. Users are required to assign a value to this field, choosing one of the policy types that define the root of trust.
+	// "PublicKey" indicates that the policy relies on a PGP publicKey and may optionally use a Rekor verification.
+	// "FulcioCAWithRekor" indicates that the policy is based on the Fulcio certification and incorporates a Rekor verification.
+	// +unionDiscriminator
+	// +kubebuilder:validation:Required
+	PolicyType PolicyType `json:"policyType"`
+	// publicKey defines the root of trust based on a PGP public key.
+	// +optional
+	PublicKey *PublicKey `json:"publicKey,omitempty"`
+	// fulcioCAWithRekor defines the root of trust based on the Fulcio certificate and the Rekor public key.
+	// For more information about Fulcio and Rekor, please refer to the document at:
+	// https://github.com/sigstore/fulcio and https://github.com/sigstore/rekor
+	// +optional
+	FulcioCAWithRekor *FulcioCAWithRekor `json:"fulcioCAWithRekor,omitempty"`
+}
+
+// +kubebuilder:validation:Enum=PublicKey;FulcioCAWithRekor
+type PolicyType string
+
+const (
+	PublicKeyRootOfTrust         PolicyType = "PublicKey"
+	FulcioCAWithRekorRootOfTrust PolicyType = "FulcioCAWithRekor"
+)
+
+// PublicKey defines the root of trust based on a PGP public key.
+type PublicKey struct {
+	// keyData contains inline base64 encoded data of the public key.
+	// KeyData must be at most 8192 characters.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MaxLength=8192
+	KeyData string `json:"keyData"`
+	// rekorKeyData contains inline base64 data of the Rekor public key.
+	// rekorKeyData must be at most 8192 characters.
+	// +optional
+	// +kubebuilder:validation:MaxLength=8192
+	RekorKeyData string `json:"rekorKeyData,omitempty"`
+}
+
+// FulcioCAWithRekor defines the root of trust based on the Fulcio certificate and the Rekor public key.
+type FulcioCAWithRekor struct {
+	// fulcioCAData contains inline base64 data for the fulcio CA certificate.
+	// fulcioCAData must be at most 8192 characters.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MaxLength=8192
+	FulcioCAData string `json:"fulcioCAData"`
+	// rekorKeyData contains inline base64 data of the Rekor public key.
+	// rekorKeyData must be at most 8192 characters.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MaxLength=8192
+	RekorKeyData string `json:"rekorKeyData"`
+	// fulcioSubject specifies OIDC issuer and the email of the Fulcio authentication configuration.
+	// Required if rootOfTrust is based on Fulcio.
```

Comment:
Where is `rootOfTrust`? That's the name of the parent discrminated union right? I think we don't need to have this line since it's obvious that the field is required because we are already in the fulcio struct